Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,493 @@
+            /// </returns>
+            public override AstVisitAction VisitScriptBlockExpression(ScriptBlockExpressionAst scriptBlockExpressionAst)
+            {
+                if (!(scriptBlockExpressionAst.Parent is CommandAst commandAst))
+                {
+                    return AstVisitAction.Continue;
+                }
+
+                string cmdName = commandAst.GetCommandName();
+                if (cmdName == null)
+                {
+                    // Skip for situations where command name cannot be resolved like `& $commandName -ComputerName -ScriptBlock { $foo }` 
+                    return AstVisitAction.Continue;
+                }
+                
+                // We need this information, because some cmdlets can have more than one ScriptBlock parameter
+                var scriptBlockParameterAst = commandAst.CommandElements[
+                        commandAst.CommandElements.IndexOf(scriptBlockExpressionAst) - 1] as CommandParameterAst;
+
+                if (IsInlineScriptBlock(cmdName) || 
+                    IsJobScriptBlock(cmdName, scriptBlockParameterAst) ||
+                    IsForeachScriptBlock(cmdName, scriptBlockParameterAst) ||
+                    IsInvokeCommandComputerScriptBlock(cmdName, commandAst) ||
+                    IsDSCScriptResource(cmdName, commandAst))
+                {
+                    AnalyzeScriptBlock(scriptBlockExpressionAst);
+                    return AstVisitAction.SkipChildren;
+                }
+
+                if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+                {
+                    if (!TryGetSessionNameFromInvokeCommand(commandAst, out var sessionName))
+                    {
+                        return AstVisitAction.Continue;
+                    }
+
+                    IEnumerable<VariableExpressionAst> varsInLocalAssignments = FindVarsInAssignmentAsts(scriptBlockExpressionAst);
+                    if (varsInLocalAssignments != null)
+                    {
+                        AddAssignedVarsToSession(sessionName, varsInLocalAssignments);
+                    }
+
+                    GenerateDiagnosticRecords(
+                        FindNonAssignedNonUsingVarAsts(
+                            scriptBlockExpressionAst, 
+                            GetAssignedVarsInSession(sessionName)));
+
+                    return AstVisitAction.SkipChildren;
+                }
+                
+                return AstVisitAction.Continue;
+            }
+
+            /// <summary>
+            /// FindVarsInAssignmentAsts: Retrieves all assigned variables from an Ast:
+            /// Example: `$foo = "foo"` ==> the VariableExpressionAst for $foo is returned
+            /// </summary>
+            /// <param name="ast"></param>
+            private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)

```

Comment:
Rather than return an `IEnumerable<VariableExpressionAst>`, this could perhaps return an `IReadOnlyDictionary<string, VariableExpressionAst>`, a case-insensitive dictionary of variables by variable name, allowing the lookup later to be constant time rather than linear

Reply:
See comment for line 251. I like it!